docker/buoy/Dockerfile

```diff
@@ -1,45 +1,74 @@
-ENV RMW_IMPLEMENTATION rmw_cyclonedds_cpp
-
 # Necessary tools
 RUN apt update \
  && apt install -y \
         apt-utils \
+        vim \
         wget
 
-# Using non-official Gazebo + ROS combination, set it explicitly
-ENV GZ_VERSION garden
-
 # Garden only has nightlies for now
 RUN /bin/sh -c 'echo "deb http://packages.osrfoundation.org/gazebo/ubuntu-stable `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-stable.list' \
  && /bin/sh -c 'echo "deb http://packages.osrfoundation.org/gazebo/ubuntu-nightly `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-nightly.list' \
  && /bin/sh -c 'wget http://packages.osrfoundation.org/gazebo.key -O - | apt-key add -' \
- && apt-get update \
- && apt-get install -y gz-garden
+ && apt update \
+ && apt install -y \
+        ros-humble-rmw-cyclonedds-cpp \
+        gz-garden \
+ && apt clean
+
+# For timing in tests, need to use cyclonedds for ROS 2 rather than default
+# rmw provider
+ENV RMW_IMPLEMENTATION rmw_cyclonedds_cpp
+# Using non-official Gazebo + ROS combination, set it explicitly
+ENV GZ_VERSION garden
+
+# Add a user with the same user_id as the user outside the container
+# Requires a docker build argument `user_id`
+ARG user_id
+ENV USERNAME developer
+RUN useradd -U --uid ${user_id} -ms /bin/bash $USERNAME \
+ && echo "$USERNAME:$USERNAME" | chpasswd \
+ && adduser $USERNAME sudo \
+ && echo "$USERNAME ALL=NOPASSWD: ALL" >> /etc/sudoers.d/$USERNAME
+
+# Commands below run as the developer user
+USER $USERNAME
+
+# When running a container start in the developer's home folder
+WORKDIR /home/$USERNAME
 
 # Create project directory and import packages
-RUN mkdir -p /tmp/buoy_ws/src \
-  && cd /tmp/buoy_ws/src/ \
-  && wget https://raw.githubusercontent.com/osrf/buoy_entrypoint/main/buoy_all.yaml \
-  && vcs import < buoy_all.yaml
+ENV BUOY_WS /home/$USERNAME/buoy_ws
+RUN mkdir -p ${BUOY_WS}/src \
+ && cd ${BUOY_WS}/src/ \
+ && wget https://raw.githubusercontent.com/osrf/buoy_entrypoint/main/buoy_all.yaml \
+ && vcs import < buoy_all.yaml
 
-# Install rosdep dependencies - this installs Gazebo and other packages
+# Install rosdep dependencies
 RUN sudo apt update \
-  && cd /tmp/buoy_ws \
-  && rosdep update \
-  && rosdep install --from-paths src --ignore-src -r -y -i \
-  && rm -rf /var/lib/apt/lists/* \
-  && apt clean
+ && cd ${BUOY_WS} \
+ && rosdep update \
+ && rosdep install --from-paths src --ignore-src -r -y -i \
+ && sudo rm -rf /var/lib/apt/lists/* \
+ && sudo apt clean
 
 # Build the project
 RUN /bin/bash -c 'source /opt/ros/${ROS_DISTRO}/setup.bash \
-  && cd /tmp/buoy_ws \
+  && cd ${BUOY_WS} \
   && colcon build'
 
-ENTRYPOINT ["/bin/bash" , "-c" , "source /tmp/buoy_ws/install/setup.bash && /bin/bash"]
+# Add quick access scripts
+ENV SETUP_SH /home/$USERNAME/setup.bash
+RUN touch ${SETUP_SH} \
+ && chmod 755 ${SETUP_SH} \
+ && echo ". ${BUOY_WS}/install/setup.bash" >> ${SETUP_SH}
+ENV RUN_SH /home/$USERNAME/run_simulation.bash
```

Comment:
What's this file for? For running the simulation by default when going for the non-dev option?

Also, might need to update readme and install tutorial as well with the new instruction.

Reply:
This is a shortcut for people to run the simulation. Someone pulling the Docker image from DockerHub will not have ready access to the README, and they'd have to dig through the files in the image. Having a shortcut enables them to quickly see what the Dockerfile does.

I didn't put it as an entrypoint, because we expect the users to be controls researchers, so they'll probably be developers writing plugins, and might not want the simulator to run every time they go into the container.

README - yeah will need to update.

Reply:
Updated in 7ce8715